### PR TITLE
Reuse edit mode add widget screen in JSON editor

### PR DESCRIPTION
**Pull request by rogerl50**
Change 'add widget' code in jsonedit.js to simply show the add overlay.

Modify addWidgetLocal to return the id of the added widget, and use that id in various places in editmode.js to properly configure the JSON editor when the add overlay is done.

Edit: Updated PR description. This PR does several things, all of which are in support of the main function, which is allowing the 'add widget' code in the JSON editor to use the existing widget add overlay in the basic editor.

In jsonedit.js:
- Change 'add widget' button to just call  'showOverlay("addOverlay")'
- Add a few comments here and there

In editmode.js:
- Have 'addWidgetLocal' return the id of the added widget. (Also, modified most uses of addWidgetLocal in the codebase to use this new feature as appropriate [deckeditor.js and widget.js along with editmode.js].)
- Add a new routine, "overlayDone", to hide the add overlay and, if it was invoked from the JSON editor, make sure that the id of the new widget is highlighted in the editor. This routine is now called throughout instead of "showOverlay()".
- Modified populateWidgetOverlay to add widgets in the Interactive panel in the same order in which they appear in the panel. At the same time, laid out the panel differently, purely for reasons of appearance.
- Make the event handler for #addCustomWidgetOverlay simply add an empty basic widget instead of a basic widget containing the Ace of spades.

In room.html and editmode.css:
- Placed the three buttons on the "Custom" panel on separate lines, and renamed "Add custom widget" to "Add basic widget".

### CODE REVIEW

**Comment by rogerl50** [comment]
> > * I think there's no way to add an empty deck now. But having the one with holder and recall button is probably fine. I usually create those anyway.
> 
> Maybe the Add custom widget button should default to a null type and a random ID (e.g. {"type":, null, "id": "rand"} instead of {"image": "/i/cards-default/AS.svg"}. Then it would work pretty much like what we have now for manually adding an empty deck. Just change the type, and possibly the ID.

Seems like the real problem is the fact that the basic editor does not have a way to add an empty deck. Perhaps it would be cleaner to add a new button to the basic editor, "Add empty deck", that replicates the current functionality of adding an empty deck in the JSON editor.

**Comment by bjalder26** [comment]
> I think you wanted to answer and not edit my comment. But yeah, that sounds like a good idea.

Sorry, that was me... I usually catch it when I make that mistake.

**Comment by bjalder26** [comment]
> So is there a consensus? I see a couple of alternatives:
> 
> 1. Leave things the way they are with this PR,
> 2. Make it so that clicking the "add custom widget" button from the JSON editor inserts an empty widget, but doing so from the basic editor does what it does now,
> 3. Change the "add custom widget" button function so that it inserts a basic widget in both editors, or
> 4. Don't implement this PR.

I like this PR.  I was thinking #2, and would be plenty happy with 2, but if you're saying option 3 is that the add custom widget button would insert a basic widget from the older edit overlay when the JSON editor isn't open, but from the JSON editor when the JSON editor IS open, then maybe that is even better.

**Comment by rogerl50** [comment]
> I like this PR. I was thinking 2, and would be plenty happy with 2, but if you're saying option 3 is that the add custom widget button would insert a basic widget from the older edit overlay when the JSON editor isn't open, but from the JSON editor when the JSON editor IS open, then maybe that is even better.

Yes. Option 3 means that the "add custom widget" button would behave exactly the same way when invoked from either editor - it would insert a basic widget. I'd probably rename the button to "add basic widget".

**Comment by bjalder26** [comment]
> However, all of this would mean that the new widget would be invisible on the screen, so I'd also propose adding a `"text": "Basic widget"` property.

Hmm.. yeah I guess you'd either need to add some visible property or make it automatically switch to edit mode.  Personally I don't really have a preference.

File: client/js/jsonedit.js [hunk 1]
```diff
@@ -1458,9 +1462,9 @@ function jeSelect(start, end, scrollToCursor) {
 // Set the text area to the formatted version of the given text and colorize.
 function jeSet(text, dontFocus) {
   try {
-    $('#jeText').textContent = jePreProcessText(JSON.stringify(jePreProcessObject(JSON.parse(text)), null, '  '));
+    jeGetEditorContent(jePreProcessText(JSON.stringify(jePreProcessObject(JSON.parse(text)), null, '  ')));
   } catch(e) {
-    $('#jeText').textContent = text;
+    jeGetEditorContent(text);
```
Comment (rogerl50) [inline client/js/jsonedit.js:1467]: Both of these s/b `jeSetEditorContent`.

Comment (ArnoldSmith86) [reply client/js/jsonedit.js]: Oh, damn, thanks!

**Review by ArnoldSmith86 — APPROVED** [review]
The refactored function looks good. Should have been done a long time ago. This should fix your overlay problem although I didn't test it.

We'll look at the calls needed to get the context right in another PR. That's pretty much an unrelated issue and if you found a combination of functions that works, fine.

I'd say go for it.

File: client/js/jsonedit.js [hunk 2]
```diff
@@ -785,9 +785,8 @@ async function jeApplyChanges() {
     jeDeltaIsOurs = true;
     await jeApplyExternalChanges(completeState);
     jeStateBeforeRaw = currentStateRaw;
-    const oldState = jeStateBefore;
     jeStateBefore = currentState;
-    await updateWidget(currentState, oldState); // in editmode.js
+    await updateWidget(currentState, jeStateBefore); // in editmode.js
     jeDeltaIsOurs = false;
   }
```
Comment (rogerl50) [inline client/js/jsonedit.js:791]: @ArnoldSmith86 How was this variable unnecessary? (jeStateBefore is reset in the intervening statement.)

Comment (rogerl50) [reply client/js/jsonedit.js:792]: Reverted this change; testcafe tests now pass.

Comment (ArnoldSmith86) [reply client/js/jsonedit.js:792]: Oh. Guess it was too late yesterday.

**Pull request merged**